### PR TITLE
Adding flow types

### PR DESCRIPTION
**Pull request by Azero123**
#### Summary
Updating src/actions/posts.js to use Flow. Some changes were made to have conformity amongst the redux functions.

#### Ticket Link
mattermost/mattermost-server#9771

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit tests passed
- [x] Ran `make flow` to ensure type checking passed
- [x] Added or updated unit tests (required for all new features)
- [x] All new/modified APIs include changes to the drivers

#### Test Information
This PR was tested on: [Macbook Pro, Mac OS 10.13.6]

### CODE REVIEW

File: src/actions/posts.js [hunk 1]
```diff
@@ -165,7 +171,7 @@ export function createPost(post, files = []) {
             return {data: true};
         }
 
-        let newPost = {
+        let newPost: Object = {
```
Comment (hmhealey) [inline src/actions/posts.js]: Shouldn't this be a Post?

File: src/actions/posts.js [hunk 2]
```diff
@@ -790,7 +796,7 @@ export function getProfilesAndStatusesForPosts(postsArrayOrMap, dispatch, getSta
     return Promise.all(promises);
 }
 
-export function getNeededAtMentionedUsernames(state, posts) {
+export function getNeededAtMentionedUsernames(state: GlobalState, posts: Array<Post>): ?Set<string> {
```
Comment (hmhealey) [inline src/actions/posts.js]: Can this just return a `Set<string>` without the `?`? It doesn't look like we ever return null here

File: src/actions/posts.js [hunk 3]
```diff
@@ -808,27 +814,30 @@ export function getNeededAtMentionedUsernames(state, posts) {
 
         let match;
         while ((match = pattern.exec(post.message)) !== null) {
+            // $FlowFixMe, flow does not recognize the check of !== null in the while loop
```
Comment (hmhealey) [inline src/actions/posts.js]: We might be able to fix this by changing the loop to use a `while(true)` instead like
```javascript
while (true) {
    const match = pattern.exec(post.message);
    if (!match) {
        break;
    }

    // ...
}
```

File: src/actions/posts.js [hunk 4]
```diff
@@ -899,8 +908,8 @@ export function getNeededCustomEmojis(state, posts) {
     return customEmojisToLoad;
 }
 
-export function removePost(post) {
-    return (dispatch, getState) => {
+export function removePost(post: Post): ActionFunc {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
```
Comment (hmhealey) [inline src/actions/posts.js]: Why change this to be an async function?

File: src/types/actions.js [hunk 5]
```diff
@@ -17,12 +17,13 @@ export type GenericAction = {|
...
     currentUserId?: string,
     remove?: Function,
+    url?: string
 |};
 
 type Thunk = (DispatchFunc, GetStateFunc) => Promise<ActionResult>; // eslint-disable-line no-use-before-define
 
 type BatchAction = {
-    type: 'BATCHING_REDUCER.BATCH';
+    type: string;
```
Comment (hmhealey) [inline src/types/actions.js]: Why change this type? If I understand correctly, the previous version of this just means that the `type` field always has to have that value

File: src/types/posts.js [hunk 6]
```diff
@@ -29,12 +29,18 @@ export type PostType = 'system_add_remove' |
 
 export type PostEmbedType = 'image' | 'message_attachment' | 'opengraph';
 
+/* eslint-disable no-use-before-define */
```
Comment (hmhealey) [inline src/types/posts.js]: Why is this needed here?

File: src/types/preferences.js [hunk 7]
```diff
@@ -6,5 +6,5 @@ export type PreferenceType = {|
     category: string,
     name: string,
     user_id: string,
-    value: string
+    value?: string
```
Comment (hmhealey) [inline src/types/preferences.js]: Why is this being changed to optional?

File: src/actions/posts.js (hunk 4 above)
Comment (Azero123) [reply src/actions/posts.js]: I asked earlier in the developer community what some thought as to if we should standardize them all to be the same since they all have the behavior of returning a dispatch and getstate function and it seemed that this was the preference.

Comment (hmhealey) [reply src/actions/posts.js]: Oh, I see. I didn't realize that the `ActionFunc` required it to be async. This is fine to leave as is then, although I'm generally not a fan of adding async everywhere since it makes it harder to test this code.

**Comment by hanzei** [comment]
@Azero123 Do you have questions about the provided feedback? Looking forward to seeing this PR merged.

File: src/actions/posts.js (hunk 4 above)
Comment (sudheerDev) [reply src/actions/posts.js]: I didn't realise that we are changing them to be promises for `ActionFunc` type when this was being discussed. Rather than changing this to be `async` i think we should change the type to be either a `func` or a `promise`. Same with `return`. If we are returning `{data: true}` just for the sake of returning  then it does not add much value so we probably should change the type for return too. 
i am 0/5 on the change for return type but for changing async i am at 2/5 that we should change the type to be forgiving if it is a a simple func.

@jespino Thoughts?

Comment (jespino) [reply src/actions/posts.js]: Yes, I think is a valid usage to have sync actions, so I agree with @sudheerDev, we should change the `ActionFunc` to accept both, regular functions or async functions. about the return, I would say was the pre-existing pattern, so I wouldn't change it for now, maybe in the future, we have to rethink a lot of things in mattermost-redux.

Comment (sudheerDev) [reply src/actions/posts.js]: @Azero123 Can you change the type of `ActionFunc` to accept a func? And change back the instances of newly added async function to be sync as before?. I am good with changing return type to be `{data: true}`.
Sorry about the confusion and delay on reviewing this.

**Comment by hanzei** [comment]
Hey @Azero123,

Please note that we will close this PR after another 10 days of inactivity. If you don't object, other contributors might complete this PR for you.

**Comment by hanzei** [comment]
Closing this PR due to inactivity. Still, thank you for this contribution @Azero123 :+1:
